### PR TITLE
Create Post

### LINKED ISSUE #42: Create a Post
As an author, I would like to be able to create Posts so I can share my thoughts with the world.

**Given** a user is in the app  
**When** they select the `New Post` menu option
**Then** they should be directed to a page with a form for creating a new post

**Given** the user has entered the relevant information for a Post
**When** they click the `Save` button  
**Then** the Post should be saved to the database  
**And** the creation datetime should be automatically set to the current date and time  
**And** the post should automatically be approved  
**And** the user should be recorded as the author of the Post  
**And** the user should be redirected to the new Post's details page

The "relevant information" for a Post is

* Title
* Content
* Category
* Header Image URL (optional)
* Publication Date (optional)

### PR DESCRIPTION
**Pull request by cjgochanour**
#### Changes Made
1. Added file `PostForm` to `tabloid/client/src/components` directory.
2. Modified file `postManager` to include `addPost()` that sends a POST request to the post database.
3. Modified file `PostController` to include `Post(Post post)` that uses the data sent from `addPost()` to update the database.
4. Modified file `PostRepository` to include `AddPost(Post post)` that uses the data sent from `Post(Post post)` to build and execute the SQL query responsible for updating the database.
5. Modified file `ApplicationViews` to include routing to the new post form
6. Modified file `Header` to include a nav link to the new post form

​
#### Steps to Review
1. Checkout this branch locally.
    ```
    git fetch --all
    git checkout cg-createPost
    ```
2. Open a new Terminal tab (⌘T) and navigate to the server directory.
3. Test app functionality.
    > Log into app if not yet logged in
    > Use nav bar link to navigate to "/posts/add", or manually type in url
    > Make a new post. Optional input fields are Date and Header Image URL
    > If invalid information is added, the user will be given an error and unable to proceed
    > If valid information is added, the database will be updated with the new post and the user will be redirected to the post's detail view
4. View code file.
    > Confirm file modifications are present as indicated above.
    > Confirm no unused code or extraneous comments exist.

Closes #42

### CODE REVIEW

**Review by user1hunter — COMMENTED** [review]
When adding a post, everything works well as long as a date is added; however, when the date input is left empty, the post will not appear in the post list and the publish date in the database is set to NULL.

**Review by user1hunter — APPROVED** [review]

**Comment by cjgochanour** [comment]
> When adding a post, everything works well as long as a date is added; however, when the date input is left empty, the post will not appear in the post list and the publish date in the database is set to NULL.

This is intentional. Users may make posts without a published date time - these will not show in the post list. Users may also make a post with a future publish date time - these will not show in the post list until that date time has occurred IRL.

File: Tabloid/client/src/components/ApplicationViews.js [hunk 1]
```diff
@@ -9,52 +9,49 @@ import PostList from "./PostList.js";
...
+        <main>
+            <Switch>
+                <Route path="/" exact>
+                    {isLoggedIn ? <Hello /> : <Redirect to="/login" />}
+                </Route>
+
+                <Route path="/posts" exact>
+                    {isLoggedIn ? <PostList /> : <Redirect to="/login" />}
+                </Route>
+                <Route path="/posts/add" exact>
```
Comment (heymonicakay) [inline Tabloid/client/src/components/ApplicationViews.js]: Does this route need to be exact?

File: Tabloid/client/src/components/PostForm.js [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    const getCategories = () => {
+        getAllCategories().then((cats) => setCategories(cats));
+    };
+
+    const makePost = (e) => {
+        e.preventDefault();
+        const post = {
+            title,
+            content,
+            ImageLocation: header,
```
Comment (heymonicakay) [inline Tabloid/client/src/components/PostForm.js]: Watch your casing here.. Client side should all be camel case (when working in React especially).

File: Tabloid/client/src/components/PostForm.js [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+                <FormGroup>
+                    <Label for="dateInput">Date</Label>
+                    <Input
+                        id="dateInput"
+                        name="date"
+                        type="datetime-local"
+                        onInput={(e) => setPublicationDate(e.target.value)}
+                    />
+                </FormGroup>
+                <Button onClick={makePost}>Submit</Button>
```
Comment (heymonicakay) [inline Tabloid/client/src/components/PostForm.js]: Consider adding some kind of form validation to disable the button until the required form inputs are complete.

**Review by heymonicakay — CHANGES_REQUESTED** [review]
Please see my comments

File: Tabloid/client/src/components/PostForm.js [hunk 4]
```diff
@@ -87,7 +87,9 @@ export const PostForm = () => {
                         onInput={(e) => setPublicationDate(e.target.value)}
                     />
                 </FormGroup>
-                <Button onClick={makePost}>Submit</Button>
+                <Button disabled={!title || !content || !chosenCategory || chosenCategory === 0} onClick={makePost}>
```
Comment (heymonicakay) [inline Tabloid/client/src/components/PostForm.js:90]: Is the category nullable? if so, perhaps the user should be allowed to leave that one blank?

**Review by heymonicakay — COMMENTED** [review]
See my question...

File: Tabloid/client/src/components/PostForm.js (hunk 4 above)
Comment (cjgochanour) [reply Tabloid/client/src/components/PostForm.js:90]: Category is not nullable. The Post table has CategoryId as a NOT NULL Int, and #42 has category listed as non-optional.

Comment (heymonicakay) [reply Tabloid/client/src/components/PostForm.js:90]: Gotcha! Thanks for clarifying! You're all set then.

**Review by heymonicakay — APPROVED** [review]

**Pull request merged**